spark/src/main/scala/org/apache/spark/sql/delta/commands/VacuumCommand.scala

```diff
@@ -187,15 +190,65 @@ object VacuumCommand extends VacuumCommandImpl with Serializable {
       val partitionColumns = snapshot.metadata.partitionSchema.fieldNames
       val parallelism = spark.sessionState.conf.parallelPartitionDiscoveryParallelism
 
-      val allFilesAndDirs = DeltaFileOperations.recursiveListDirs(
+      val allFilesAndDirs = if (partitionColumns.nonEmpty) {
+        val dirsInFs = partitionColumns.foldLeft(Seq(basePath)) {
+          case (parentPaths, partition) =>
+            DeltaFileOperations.listDirs(
+              spark,
+              hadoopConf,
+              parentPaths,
+              dirFilter = p => DeltaTableUtils.isHiddenDirectory(partitionColumns, p),
+              fileFilter = DeltaTableUtils.isHiddenDirectory(partitionColumns, _),
+              fileListingParallelism = Option(parallelism))
+        }
+        val vacuumFolders = if (userPartitionPredicates.nonEmpty) {
+          val filters = userPartitionPredicates.flatMap { predicate =>
+            val predicates = parsePredicates(spark, predicate)
```

Comment:
Should parsing of the predicates come _before_ retrieving all of the directories from the fs? Right now this looks like if you have a problem with the predicate clause, it won't raise an error until after it has walked the directory tree.

Reply:
@watfordkcf Thank you for the suggestion, I have updated the logic to generate valid filter expressions at the beginning. For non-partitioned tables, using a filter would trigger an Exception.